File: ftc_app-3.5/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robots/Robot.java
package org.firstinspires.ftc.teamcode.robots;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

public abstract class Robot {
	public LinearOpMode linearOpMode;

	public enum AllianceColor {
		RED, BLUE, UNKNOWN
	}

    public abstract void setup(LinearOpMode linearOpMode);

    public void turn(double power, int degrees) {
	    if(degrees > 0) {
		    double targetHeading = getGyroHeading() + degrees;
		    setTurnPower(Math.abs(power));
		    while(getGyroHeading() < targetHeading && linearOpMode.opModeIsActive()) {
			    linearOpMode.telemetry.addData("Heading", getGyroHeading());
			    linearOpMode.telemetry.update();
			    linearOpMode.idle();
		    }
	    } else if(degrees < 0) {
		    double targetHeading = getGyroHeading() + degrees;
		    setTurnPower(-Math.abs(power));
		    while(getGyroHeading() > targetHeading && linearOpMode.opModeIsActive()) {
			    linearOpMode.telemetry.addData("-Heading", getGyroHeading());
			    linearOpMode.telemetry.update();
		    	linearOpMode.idle();
		    }
	    }

	    stopAllDriveMotors();
    }

    public abstract void setDrivePower(double power);
    public abstract void setTurnPower(double power);
    public abstract void stopAllDriveMotors();
    public abstract double getGyroHeading();
}

File: ftc_app-3.5/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robots/Robot.kt
package org.firstinspires.ftc.teamcode.robots

import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.hardware.HardwareMap

/**
 * A base class for creating a robot of any type.
 *
 * @author FTC Team 5037 gotrobot?
 */
abstract class Robot {

    lateinit var opMode: OpMode

    /**
     * Sets up the hardware needed in order to use the robot.
     * @param hardwareMap A HardwareMap object. Usually provided by opMode.hardwareMap.
     */
    abstract fun setup(hardwareMap: HardwareMap)

    /**
     * Sets the power that the robot should drive forward or backwards.
     * @param power The power that the robot should drive at.
     */
    abstract fun setDrivePower(power: Double)

    /**
     * Sets the power that the robot should turn.
     * @param power The power that the robot should turn at.
     */
    abstract fun setTurnPower(power: Double)

    /**
     * Sets the power of the drive motors to 0.
     */
    abstract fun stopAllDriveMotors()

    /**
     * Turns the robot to the specified angle.
     * @param power The power that the robot should turn at.
     * @param degrees the degrees the robot should move at.
     */
    abstract fun turn(power: Double, degrees: Double)
}
